specs/proposals/diagnostic-experience.md

```diff
@@ -0,0 +1,249 @@
+
+- **Purpose**: Debug logs are specifically aimed to facilitate `notation` developers to diagnose `notation` tool itself. They contain detailed technical information that is useful for troubleshooting problems.
+- **Target users**: Primarily intended for developers or technical users who are trying to understand the inner workings of the code and identify the root cause of a possible issue with the tool itself.
+- **Content**: Debug logs focus on providing context needed to troubleshoot issues, like variable values, execution paths, error stack traces, and internal states of the application.
+- **Level of Detail**: Extremely detailed, providing insights into the application's internal workings and logic, often including low-level details that are essential for debugging.
+
+Currently, the verbose output of `notation` prints INFO level of logs, which is overlapped with debug logs. This is duplicated information for users.
+
+### Output
+
+There are four types of output in `notation` CLI:
+
+- **Status output**: such as operation progress information and command execution result.
+- **Metadata output**: showing what has been executed in specified format, such as JSON, text.
+- **Content output**: it is to output the raw data obtained from the remote registry server or file system, such as the generated signature file.
+- **Error output**: error messages are expected to be helpful to troubleshoot where the user has done something wrong and the program is guiding them in the right direction.
+
+The target users of these types of output are general users. 
+
+## Proposals
+
+### Common Conventions
+
+By defining the common conventions, it helps `notation` print out clear and analyzable debug logs.
+
+- Timestamp Each Log Entry with precise timing: Ensure each log entry has a precise timestamp to trace the sequence of events accurately. `notation` SHOULD use the [Nanoseconds](https://pkg.go.dev/time#Duration.Nanoseconds) precision to print the timestamp in the first field of each line. Example: `[2024-08-02 23:56:02.6738192Z] `
+- Avoid logging sensitive information for privacy and security requirement: Abstain from logging sensitive information such as passwords, personal data, or authentication tokens. Example: `[2024-08-02 23:56:02.7338192Z] Attempting to authenticate user [UserID: usr123]` (`notation` SHOULD exclude authentication token and password information).
+
+### Enhancements
+
+Here are the proposals for `notation` diagnostic experience enhancements: 
+
+- Deprecate the `--verbose` flag but keep `--debug` flag to avoid ambiguity and duplicated INFO level logs in two outputs. It is reasonable to continue using `--debug` to enable logs with different levels as it is in `notation`.
+- Add two empty lines as the separator between each request and response for readability.
```

Comment:
If we can use colored output, the log section can be easily separated by log level with color. In that case, are the two empty lines still necessary?

Reply:
In case the color-coded log level may work in some cases, such as Windows PowerShell terminal